single_kernel_mongo/managers/sharding.py

```diff
@@ -0,0 +1,807 @@
+            raise NonDeferrableFailedHookChecksError("relation is not feasible")
+        if not self.charm.unit.is_leader():
+            raise NonDeferrableFailedHookChecksError
+        if (
+            revision_mismatch_status
+            := self.dependent.cluster_version_checker.get_cluster_mismatched_revision_status()
+        ):
+            self.charm.status_manager.set_and_share_status(revision_mismatch_status)
+            raise DeferrableFailedHookChecksError("Mismatched versions in the cluster")
+        if not self.state.is_role(MongoDBRoles.CONFIG_SERVER):
+            raise NonDeferrableFailedHookChecksError("is only executed by config-server")
+
+    def assert_pass_hook_checks(self, relation: Relation, leaving: bool = False) -> None:
+        """Runs pre hooks checks and raises the appropriate error if it fails.
+
+        Raises:
+            NonDeferrableFailedHookChecksError, DeferrableFailedHookChecksError
+        """
+        self.assert_pass_sanity_hook_checks()
+
+        pbm_status = self.dependent.backup_manager.get_status()
+        if isinstance(pbm_status, MaintenanceStatus):
+            raise DeferrableFailedHookChecksError(
+                "Cannot add/remove shards while a backup/restore is in progress."
+            )
+
+        if self.state.upgrade_in_progress:
+            logger.warning(
+                "Adding/Removing shards is not supported during an upgrade. The charm may be in a broken, unrecoverable state"
+            )
+            if not leaving:
+                raise DeferrableFailedHookChecksError("Upgrade is in progress")
+            if self.state.has_departed_run(relation.id):
+                raise DeferrableFailedHookChecksError(
+                    "must wait for relation departed hook to decide if relation should be removed"
+                )
+            self.dependent.assert_proceed_on_broken_event(relation)
+
+    def update_credentials(self, key: str, value: str) -> None:
+        """Sends new credentials for a new key value pair across all shards."""
+        for relation in self.state.config_server_relation:
+            if self.data_interface.fetch_relation_field(relation.id, "database") is None:
+                logger.info(f"Database Requested event has not run yet for relation {relation.id}")
+                continue
+            self.data_interface.update_relation_data(relation.id, {key: value})
+
+    def update_mongos_hosts(self) -> None:
+        """Updates the hosts for mongos on the relation data."""
+        for relation in self.state.config_server_relation:
+            self.data_interface.update_relation_data(
+                relation.id,
+                {
+                    AppShardingComponentKeys.HOST.value: json.dumps(
+                        sorted(self.state.internal_hosts)
+                    )
+                },
+            )
+
+    def update_ca_secret(self, new_ca: str | None) -> None:
```

Comment:
Is this still needed? I noticed that `charm_state` now has this function and that this function was removed in the mongos-config-server manager

Reply:
Yeah this needs to be removed :)